### ISSUE TITLE
[BUG] Very high CPU usage on Oracle Linux 8.6 with Kernel 5.15

### ISSUE THREAD

**Issue by msr-financial-com** [rancher/rancher#38816]
**Rancher Server Setup**
- Rancher version: v2.6.8
- Installation option : Helm Chart
   - RKE1
   - Kubernetes: v1.24.4

**Information about the Cluster**
- Kubernetes version: v1.24.4
- Cluster Type (Local/Downstream): Downstream
   -  Custom

**User Information**
- What is the role of the user logged in? Admin/Cluster

**Describe the bug**
When creating a Custom downstream cluster with Kubernetes v1.24.x and using servers with Oracle Linux 8.6 and the kernel-uek-core 5.15, the CPU usage goes abnormally high. This issue does not occur with the Kernel kernel-core 4.18 and Kubernetes 1.24.x. No extra pods have to be deployed. It will occur immediately after the node is registered into the cluster

**To Reproduce**
- Create a downstream cluster via Rancher UI and use Kubernetes v1.24 or add node to an v1.24 cluster
- Add serves with Oracle Linux 8.6 installed and Kernel-uek-core 5.15
- Display cpu usage("top")

**Result**
CPU usage goes abnormally high. While only minimum of pods are deployed
Because of this, pods are getting killed because of a shortage on CPU. 

**Expected Result**
Normal CPU usage.

**Screenshots**
![Number1](https://user-images.githubusercontent.com/73880028/188661390-3a876e49-81c1-415a-a1ad-475d0e3a8eac.png)
![Number5](https://user-images.githubusercontent.com/73880028/188662281-a686e49e-1997-47da-886c-71c6e139a25d.png)
![Number2](https://user-images.githubusercontent.com/73880028/188661417-fc2ed4f8-8a61-4f8d-9ff8-b0c3b3bdd702.png)
![number4](https://user-images.githubusercontent.com/73880028/188661610-8c51dc30-15dc-475f-94bb-c10a6c18f47b.png)
![number3](https://user-images.githubusercontent.com/73880028/188661627-3fd90ec2-8033-4b1a-83f6-f3319d9edac1.gif)
**Additional context**

**Comment by cite** [comment]
Can confirm. This is affecting RKE1 clusters, and it doesn't matter whether it's an upgrade or a new installation. Depending on the number of pods, even when there are no state changes, `dockerd` will consume several 10s of CPU cores (I've seen CPU loads as high as 5000%).

The affected 5.15 kernel is called [Unbreakable Enterprise Kernel 7](https://docs.oracle.com/en/operating-systems/uek/7/relnotes7.0/) by Oracle.

**Comment by cite** [comment]
It seems this is not caused by OEL or it's kernel, since the same symptoms can be reproduced on a CentOS 7 worker node. Is it possible that the fix for https://github.com/rancher/rke/issues/2938 didn't make it into Rancher 2.6.8 yet?

**Comment by jiaqiluo** [comment]
This is a known issue: https://github.com/rancher/rancher/issues/38214  where we see the unexpected high memory usage of dockerd on OL 8.4 and 8.6 on k8s 1.24

**Comment by cite** [comment]
@jiaqiluo I don't believe this to be the same issue. In our setup, we don't see any if the error messages your linked report contains, and neither is memory consumption of the Docker daemon increased (see screenshots of the initial reporter).

**Comment by cite** [comment]
@jiaqiluo Would it help if we were able to recreate this problem with CentOS 7? If so, I'll setup a new worker node and join it.

**Comment by jiaqiluo** [comment]
Hi @cite , it would be great if you could try it on CentOS 7 with RKE1 k8s 1.24.x?  Can you also share the docker version on the node?

**Comment by cite** [comment]
@jiaqiluo This is not easy to pin down. I deployed a completely new RKE1 cluster with Rancher 2.6.8, two nodes OEL 8.6 with the default 4.18 kernel, one node with CentOS 7.9 with it's default 3.18 kernel.

```
# uname -r; rpm -q containerd.io docker-ce
3.10.0-1160.76.1.el7.x86_64
containerd.io-1.6.6-3.1.el7.x86_64
docker-ce-20.10.17-3.el7.x86_64
```
After deploying a cluster with version 1.23.10-rancher1 and enabling `cri-dockerd`, and deploying 400 nginx containers, I immediately noticed very bad metrics performance on all nodes:

```# time curl -sk -H "Authorization: Bearer $b" 'https://localhost:10250/stats/summary?only_cpu_and_memory=true' >/dev/null

real	0m4.728s
user	0m0.050s
sys	0m0.061s
```

There was no difference between OEL 8.6 and CentOS 7.9 workers. Furthermore, the `kubelet` logs showed errors like that:

```
E0909 10:09:37.686805   11277 remote_runtime.go:1134] "ReopenContainerLog from runtime service failed" err="rpc error: code = Unknown desc = docker does not support reopening container log files" containerID="80faaaf76cf140f5fe00b9615face7d1c38fc9fad12f85cbf565ee318252d93b"
E0909 10:09:37.686836   11277 container_log_manager.go:244] "Container log doesn't exist, reopen container log failed" err="rpc error: code = Unknown desc = docker does not support reopening container log files" containerID="80faaaf76cf140f5fe00b9615face7d1c38fc9fad12f85cbf565ee318252d93b" path="/var/log/pods/default_echo-75f946c956-tq9jl_20de94b8-9589-4a3f-a751-d41759751e60/echo/0.log"
time="2022-09-09T10:09:39Z" level=error msg="Cannot connect to the Docker daemon at unix:///var/run/docker.sock. Is the docker daemon running? Failed to get stats from container 964410c5c8e039140b36ded75cd83500f168cfb65d023b8f33753d513f0ba19d"
[...repeated several 10s of times...]
E0909 10:09:39.066200   11277 remote_runtime.go:351] "ListPodSandbox with filter from runtime service failed" err="rpc error: code = Unknown desc = Cannot connect to the Docker daemon at unix:///var/run/docker.sock. Is the docker daemon running?" filter="nil"
E0909 10:09:39.066233   11277 kuberuntime_sandbox.go:292] "Failed to list pod sandboxes" err="rpc error: code = Unknown desc = Cannot connect to the Docker daemon at unix:///var/run/docker.sock. Is the docker daemon running?"
E0909 10:09:39.066244   11277 generic.go:205] "GenericPLEG: Unable to retrieve pods" err="rpc error: code = Unknown desc = Cannot connect to the Docker daemon at unix:///var/run/docker.sock. Is the docker daemon running?"
time="2022-09-09T10:09:41Z" level=info msg="Using CNI configuration file /etc/cni/net.d/10-canal.conflist"
time="2022-09-09T10:09:46Z" level=info msg="Using CNI configuration file /etc/cni/net.d/10-canal.conflist"
E0909 10:09:48.029518   11277 remote_runtime.go:1134] "ReopenContainerLog from runtime service failed" err="rpc error: code = Unknown desc = docker does not support reopening container log files" containerID="80faaaf76cf140f5fe00b9615face7d1c38fc9fad12f85cbf565ee318252d93b"
E0909 10:09:48.029547   11277 container_log_manager.go:244] "Container log doesn't exist, reopen container log failed" err="rpc error: code = Unknown desc = docker does not support reopening container log files" containerID="80faaaf76cf140f5fe00b9615face7d1c38fc9fad12f85cbf565ee318252d93b" path="/var/log/pods/default_echo-75f946c956-tq9jl_20de94b8-9589-4a3f-a751-d41759751e60/echo/0.log"
```

Again, there was no difference between the CentOS 7.9 and OEL 8.6 workers. The metrics performance is suprising, because https://github.com/rancher/rke/issues/2938 states that the version of `cri-dockerd` as shipped with Rancher is 0.2.4, which should include https://github.com/Mirantis/cri-dockerd/pull/38.

For all nodes, memory consumption of  `dockerd` was increased by about 600% compared to normal. Also, CPU usage for `dockerd` was about 10x what we normally see on an idle cluster - with CentOS 7.9 faring about 2x better than OEL 8.6.

When disabling `cri-dockerd`, the logfiles immediately went back to normal, CPU consumption went back to normal, and memory consumption didn't increase anymore after restarting `dockerd`. Also, metrics are available about 100x faster:

```
# time curl -sk -H "Authorization: Bearer $b" 'https://localhost:10250/stats/summary?only_cpu_and_memory=true' >/dev/null

real	0m0.116s
user	0m0.045s
sys	0m0.058s
```

Upgrading the cluster to 1.24.4 didn't make any difference to the situation of running 1.23.10 with `cri-dockerd` enabled.

Looking at the code changes for https://github.com/Mirantis/cri-dockerd/pull/38, it seems that metrics are available faster in that case because collecting them is parallelized by distributing the work to several goroutines. Is it possible that with a large number of pods, the parallel queries overwhelm `dockerd` (or hit `somaxconn` on it's unix domain socket)?

Conclusion: Enabling `cri-dockerd` seems to have a larger impact than the worker nodes' OS. However, there is a definitive degradation on OEL 8.6 when compared to CentOS 7.9.

Do you want me to perform other tests or ship more data?

**Comment by chfrank-cgn** [comment]
I've observed the same behavior on Ubuntu 20.04 LTS (RKE 1.24.4 from Rancher 2.6.8)

**Comment by chfrank-cgn** [comment]
Same issue, independent of whether I upgrade from 1.23.10 to 1.24.4 or install 1.24.4 right away - CPU usage on the nodes goes through the roof, with dockerd continuously using 290 to 310%.

Stopping kube-metrics-server and Prometheus ease the situation somewhat, but CPU usage remains high.

Right on the time of the upgrade, error messages appear in metrics server:

`E0911 17:00:10.466866       1 scraper.go:140] "Failed to scrape node" err="Get \"https://10.240.0.5:10250/metrics/resource\": dial tcp 10.240.0.5:10250: connect: connection refused" node="rke-f51ad0-0"
E0911 17:00:38.955105       1 scraper.go:140] "Failed to scrape node" err="Get \"https://10.240.0.5:10250/metrics/resource\": context deadline exceeded" node="rke-f51ad0-0"`

**Comment by chfrank-cgn** [comment]
![Screenshot_2022-09-11_19-06-33](https://user-images.githubusercontent.com/42704306/189541386-dcf10c73-7c38-4beb-a4b9-673dabb26a06.png)

**Comment by thaneunsoo** [comment]
I tried to reproduce on 2.6.8

**Reproduction Steps**
Fresh
1. Create custom rke downstream cluster with 1.24 k8s version (cri-dockerd is enabled by default)
2. Install monitoring on the cluster
3. Verify CPU usage using Grafana
4. Verify CPU usage by ssh into control plane node and running `top` command

Upgrade
1. Create custom rke downstream cluster with 1.23 k8s version using Ubuntu 20.04 (enable cri-dockerd)
2. Install monitoring on the cluster
3. Verify CPU usage using Grafana
4. Verify CPU usage by ssh into control plane node and running `top` command
5. Upgrade k8s version to 1.24
6. Repeat steps 3-4

**Result**
I was not able to reproduce the behavior seen in this ticket

**Comment by cite** [comment]
@thaneunsoo Were you able to reproduce metrics timings (with `curl`)? Why did you `ssh` into a Control Plane node and not a worker node that hosts pods? Also, your reproduction steps don't include generating a large number of pods.

**Comment by chfrank-cgn** [comment]
Next step, updating from the Rancher UI to v1.24.4-rancher1-1

CPU consumption goes way up, both in top and in Grafana:

![after-Top](https://user-images.githubusercontent.com/42704306/189966225-f370d0e4-2266-49a9-91f4-3a037416881a.png)

![after-ClusterNodes](https://user-images.githubusercontent.com/42704306/189966354-4f7cf35c-bbce-4123-bedb-ca8fe81c87a7.png)

**Comment by chfrank-cgn** [comment]
dockerd spends most of its time in path/filepath.Clean and runtime.scanobject, according to perf top

strace shows that it seems to be iterating/looping around content in/var/lib/rancher-data/local-catalogs/, repeatedly calling lgetxattr() and newfstatat(), with repeated SIGURGs in between

**Comment by danielhass** [comment]
Wanted to provide an additional reproduction of this issue. Situation during we faced this issue was reproduced by the following steps:
- Rancher v2.6.3 cluster running on k8s v1.21.8 + 2 attached custom clusters running v1.21
- Upgraded Rancher to v2.6.8
- Upgraded custom clusters to v1.24.4 (the custom clusters were running fine, no issue occurred there)
- Upgraded Rancher cluster from v1.21.8 to v1.24.4 using RKE1

After the last step the `dockerd` processes immediately starts to congest the CPU and causing ever rising system load numbers.

OS information:
```
SUSE Linux Enterprise Server 12 SP5
4.12.14-122.106-default
```

**Comment by kinarashah** [comment]
Based on my investigation, seems like the new fix for cri-dockerd is working as expected (https://github.com/Mirantis/cri-dockerd/pull/38). But I am seeing cpu spike when Native overlay diff for docker is disabled. Turning it on brings cpu usage back to normal. 

Add the following to /etc/modprobe.d/disable-overlay-redirect-dir.conf:

`options overlay metacopy=off redirect_dir=off`

Then, you can stop docker, reload the module, and restart docker to check your new status:
```
systemctl stop docker
modprobe -r overlay
modprobe overlay
systemctl start docker
```

My current thought process is this doesn't require a code fix in RKE or cri-dockerd, but need to look into this more.

**Comment by kinarashah** [comment]
P.S. @chfrank-cgn was able to turn on the native overlay and confirmed that he saw something similar.

**Comment by chfrank-cgn** [comment]
We can document the requirement for new clusters, but I believe there will be an issue with upgrades.

Imagine you have a nice downstream cluster, three control plane nodes, and ten worker nodes, happily chugging along on 1.23.10. In the Rancher UI, you select an upgrade to 1.24.4, and the cluster stops responding - reconfiguring and restarting docker on all 13 nodes is probably not what you want to do.

Is there a way to change the behavior in either RKE1 or cri_dockerd?

**Comment by cite** [comment]
@kinarashah Is there a way to verify that the setting is effective? On an CentOS 7 machine with the default kernel, we see the following output from `docker info`:

```
[...]
 Server Version: 20.10.17
 Storage Driver: overlay2
  Backing Filesystem: extfs
  Supports d_type: true
  Native Overlay Diff: true
  userxattr: false
 Logging Driver: json-file
 Cgroup Driver: systemd
 Cgroup Version: 1
[...]
```

However, metrics are still slow. The CentOS 7 kernel is 3.18, it doesn't even support the `metacopy` parameter.

For the OEL8 nodes, we tested with both the default 4.18 kernel as well as the 5.15 kernel (they accept the paramteres to the `overlay` module), same symptoms, same output of `docker info`. Tested with 1.23.10 with `cri-dockerd` explicitly enabled as well as 1.24.4.

**Comment by cite** [comment]
It seems I jumped the gun on the BTRFS statement - the docker-ce edition on RHEL distros lacks support for the BTRFS storage driver, so even with BTRFS, storage driver was still `overlay2`, just with a different backing system.

**Comment by nickvth** [comment]
After upgrading to 1.24.4 and cri-dockerd enabled we go to 200 process forks instead of 50 and much more cpu wait and high cpu load with top command.

![cpu-wait](https://user-images.githubusercontent.com/7606465/199593375-b3e42e60-3a46-478b-b242-9ead63cf0b3b.png)

![process-forks](https://user-images.githubusercontent.com/7606465/199593385-70ea91e3-4cd6-4ef6-ad93-7aec93066329.png)

**Comment by kinarashah** [comment]
@nickvth Could you provide OS info and docker output?

**Comment by nickvth** [comment]
> docker info (gives also Ubuntu version)
Client:
 Context:    default
 Debug Mode: false
 Plugins:
  app: Docker App (Docker Inc., v0.9.1-beta3)
  buildx: Docker Buildx (Docker Inc., v0.8.2-docker)
  scan: Docker Scan (Docker Inc., v0.12.0)

Server:
 Containers: 174
  Running: 146
  Paused: 0
  Stopped: 28
 Images: 61
 Server Version: 20.10.17
 Storage Driver: overlay2
  Backing Filesystem: extfs
  Supports d_type: true
  Native Overlay Diff: true
  userxattr: false
 Logging Driver: json-file
 Cgroup Driver: cgroupfs
 Cgroup Version: 1
 Plugins:
  Volume: local
  Network: bridge host ipvlan macvlan null overlay
  Log: awslogs fluentd gcplogs gelf journald json-file local logentries splunk syslog
 Swarm: inactive
 Runtimes: io.containerd.runc.v2 io.containerd.runtime.v1.linux runc
 Default Runtime: runc
 Init Binary: docker-init
 containerd version: 10c12954828e7c7c9b6e0ea9b0c02b01407d3ae1
 runc version: v1.1.2-0-ga916309
 init version: de40ad0
 Security Options:
  apparmor
  seccomp
   Profile: default
 Kernel Version: 5.4.0-131-generic
 Operating System: Ubuntu 20.04.4 LTS
 OSType: linux
 Architecture: x86_64
 CPUs: 4
 Total Memory: 15.6GiB
 Name: *****************
 ID: *****************************
 Docker Root Dir: /var/lib/docker
 Debug Mode: false
 Registry: https://index.docker.io/v1/
 Labels:
 Experimental: false
 Insecure Registries:
  127.0.0.0/8
 Live Restore Enabled: true

**Comment by cite** [comment]
@nickvth Are you allowed to say what kind of platform you are running on, e.g. bare metal, KVM, ESX, AWS, ...?

**Comment by nickvth** [comment]
@cite We run on vmware ESX.

**Comment by kinarashah** [comment]
Tried reproducing this on CentOS today. While @sowmyav27 and I saw dockerd going up to 90-100% in top, it almost instantly comes back to normal and we see no other errors in kubelet or metrics-server so I don't think we've reproduced the issue successfully yet. Docker output is responsive as well. 

Here is what I did to to reproduce: 
- Rancher v2.6.9
- Started with RKE1 1.23.12 cluster custom downstream centOS cluster: 1 cp+ep, 3 worker nodes in ec2 
- Deployed and scaled nginx deployment to 300 pods (~200 containers running on each node per docker info) 
- Monitored top and output of `'https://localhost:10250/stats/summary?only_cpu_and_memory=true'` 
- Upgraded cluster to 1.24.4 
- Monitored top and output of `'https://localhost:10250/stats/summary?only_cpu_and_memory=true'`  

`top` output for `dockerd` keeps fluctuating very fast, ranging from 6-7 to 80-90%. No errors in kubelet or metrics-server. Output of stats remains decent throughout. Highest I saw was 
```
time curl -sk --key /etc/kubernetes/ssl/kube-apiserver-key.pem --cert /etc/kubernetes/ssl/kube-apiserver.pem 'https://localhost:10250/stats/summary?only_cpu_and_memory=true' >/dev/null

real	0m0.153s
user	0m0.061s
sys	0m0.085s
``` 

docker info: 
```
Client:
 Context:    default
 Debug Mode: false
 Plugins:
  app: Docker App (Docker Inc., v0.9.1-beta3)
  buildx: Docker Buildx (Docker Inc., v0.8.2-docker)
  compose: Docker Compose (Docker Inc., v2.12.2)
  scan: Docker Scan (Docker Inc., v0.21.0)

Server:
 Containers: 222
  Running: 218
  Paused: 0
  Stopped: 4
 Images: 13
 Server Version: 20.10.17
 Storage Driver: overlay2
  Backing Filesystem: xfs
  Supports d_type: true
  Native Overlay Diff: true
  userxattr: false
 Logging Driver: json-file
 Cgroup Driver: cgroupfs
 Cgroup Version: 1
 Plugins:
  Volume: local
  Network: bridge host ipvlan macvlan null overlay
  Log: awslogs fluentd gcplogs gelf journald json-file local logentries splunk syslog
 Swarm: inactive
 Runtimes: io.containerd.runtime.v1.linux runc io.containerd.runc.v2
 Default Runtime: runc
 Init Binary: docker-init
 containerd version: 1c90a442489720eec95342e1789ee8a5e1b9536f
 runc version: v1.1.4-0-g5fd4c4d
 init version: de40ad0
 Security Options:
  seccomp
   Profile: default
 Kernel Version: 3.10.0-1160.76.1.el7.x86_64
 Operating System: CentOS Linux 7 (Core)
 OSType: linux
 Architecture: x86_64
 CPUs: 4
 Total Memory: 15.51GiB
 Name: xxx
 ID: xxxx 
 Docker Root Dir: /var/lib/docker
 Debug Mode: false
 Registry: https://index.docker.io/v1/
 Labels:
 Experimental: false
 Insecure Registries:
  127.0.0.0/8
 Live Restore Enabled: false
 ```

**Comment by kinarashah** [comment]
@sowmyav27 reproduced it on Oracle Linux 8.6 and it was the same issue as when I was debugging @chfrank-cgn's setup earlier (https://github.com/rancher/rancher/issues/38816#issuecomment-1259798220). It had different docker configuration where native diff overlay was turned off and changing that helped improve performance.. but I don't think that applies to other setups.

**Comment by cite** [comment]
@kinarashah Yes, that's about the steps we followed. We could also reproduce the problem with 1.23.x when switching to `cri-dockerd`. Our `docker info` is pretty similar to yours, but we are running on KVM hypervisors, not AWS:

```
Client:
 Context:    default
 Debug Mode: false
 Plugins:
  app: Docker App (Docker Inc., v0.9.1-beta3)
  buildx: Docker Buildx (Docker Inc., v0.8.2-docker)
  scan: Docker Scan (Docker Inc., v0.17.0)

Server:
 Containers: 143
  Running: 122
  Paused: 0
  Stopped: 21
 Images: 31
 Server Version: 20.10.17
 Storage Driver: overlay2
  Backing Filesystem: xfs
  Supports d_type: true
  Native Overlay Diff: true
  userxattr: false
 Logging Driver: json-file
 Cgroup Driver: cgroupfs
 Cgroup Version: 1
 Plugins:
  Volume: local
  Network: bridge host ipvlan macvlan null overlay
  Log: awslogs fluentd gcplogs gelf journald json-file local logentries splunk syslog
 Swarm: inactive
 Runtimes: io.containerd.runc.v2 io.containerd.runtime.v1.linux runc
 Default Runtime: runc
 Init Binary: docker-init
 containerd version: 10c12954828e7c7c9b6e0ea9b0c02b01407d3ae1
 runc version: v1.1.2-0-ga916309
 init version: de40ad0
 Security Options:
  seccomp
   Profile: default
 Kernel Version: 3.10.0-1160.76.1.el7.x86_64
 Operating System: CentOS Linux 7 (Core)
 OSType: linux
 Architecture: x86_64
 CPUs: 8
 Total Memory: 15.51GiB
 Name: xxxx
 ID: xxxx
 Docker Root Dir: /var/lib/docker
 Debug Mode: false
 HTTP Proxy: http://muc1tes-proxy-vip.example.com:3128
 HTTPS Proxy: http://muc1tes-proxy-vip.example.com:3128
 No Proxy: localhost,127.0.0.1,docker-hosted.example.com
 Registry: https://index.docker.io/v1/
 Labels:
 Experimental: false
 Insecure Registries:
  docker.example.com
  localhost
  127.0.0.1
  docker-hosted.example.com
  127.0.0.0/8
 Registry Mirrors:
  http://docker.example.com/
 Live Restore Enabled: false
```

**Comment by cite** [comment]
@kinarashah I was able to reproduce the problem on bars metal, too, so it's probably not KVM related.

**Comment by kinarashah** [comment]
Thanks much for the info @cite!

**Comment by cite** [comment]
@kinarashah @nickvth @chfrank-cgn In your tests, do you remember which of the following filesystems were on dedicated devices (i.e. not part of the root filesystem), and the filesystem type?

* /var
* /var/lib/docker
* /var/lib/kubelet

In our case, `/var/lib/docker` and `/var` are dedicated filesystems, with `/var` being ext4 and `/var/lib/docker` being XFS or ext4.

**Comment by cite** [comment]
@kinarashah I have one more data points: Onn our Rancher 2.6.9/RKE1 cluster running 1.24.6, I replaced the bundled version of `cri-dockerd` with the latest 0.26 release from the RPM (which involves patching `kubelet`'s entrypoint, I realize that voids any warranty ;-)). This release includes https://github.com/Mirantis/cri-dockerd/pull/107 and timing for stats collection goes down to about half the time:

Before:
```bash
# time curl -sk -H "Authorization: Bearer $b" 'https://localhost:10250/stats/summary?only_cpu_and_memory=true' >/dev/null

real	0m2.303s
user	0m0.013s
sys	0m0.007s
```

After:
```
# time curl -sk -H "Authorization: Bearer $b" 'https://localhost:10250/stats/summary?only_cpu_and_memory=true' >/dev/null

real	0m1.197s
user	0m0.055s
sys	0m0.072s
```

I'd like to point out that we still see this problem when switching a 1.23.x cluster to `cri-dockerd` (see for example the latest report here: https://github.com/rancher/rke/issues/2938)

**Comment by horihel** [comment]
@cite How involved is that workaround?
the updated cri-dockerd also has some other important fixes to bugs that range from annoying to breaking, so I'd be happy for pointers to usable workarounds :)

**Comment by cite** [comment]
@horihel  The actual diff to use the external running RPM version is pretty small:

```--- /opt/rke-tools/entrypoint.sh.bak	2022-11-16 09:45:04.041858278 +0000
+++ /opt/rke-tools/entrypoint.sh	2022-11-16 09:46:53.000000000 +0000
@@ -94,12 +94,7 @@ if [ "$1" = "kubelet" ]; then
           # cri-dockerd v0.2.4 requires unix socket or tcp endpoint, update old endpoint passed by rke
           CONTAINER_RUNTIME_ENDPOINT="unix://$CONTAINER_RUNTIME_ENDPOINT"
         fi
-        /opt/rke-tools/bin/cri-dockerd --network-plugin="cni" --cni-conf-dir="/etc/cni/net.d" --cni-bin-dir="/opt/cni/bin" ${RKE_KUBELET_PAUSEIMAGE} --container-runtime-endpoint=$CONTAINER_RUNTIME_ENDPOINT &
-
-        # wait for cri-dockerd to start as kubelet depends on it
-        echo "Sleeping 10 waiting for cri-dockerd to start"
-        sleep 10
-
+        
         # start kubelet
         exec "$@" --cgroup-driver=$CGROUPDRIVER $RESOLVCONF &
```

However, since `cri-dockerd` has no library dependencies, it's probably easier to just use `docker cp` to copy the updated version to the kubelet container and then restart that with `docker restart kubelet`.

**Comment by kinarashah** [comment]
@cite 

- I am finding the outputs of `time curl -sk --key /etc/kubernetes/ssl/kube-apiserver-key.pem --cert /etc kubernetes/ssl/kube-apiserver.pem 'https://localhost:10250/stats/summary?only_cpu_and_memory=true' >/dev/null` very low (and they seem to fluctuate as well): 

```
time curl -sk --key /etc/kubernetes/ssl/kube-apiserver-key.pem --cert /etc kubernetes/ssl/kube-apiserver.pem 'https://localhost:10250/stats/summary?only_cpu_and_memory=true' >/dev/null

real	0m0.018s
user	0m0.002s
sys	0m0.004s
```

- I am seeing the actual `stats/summary` and `/metrics/resource` calls give better estimate of time durations: 
```
time kubectl get --raw /api/v1/nodes/$NODE_NAME/proxy/stats/summary

real    0m27.621s
user    0m0.062s
sys     0m0.004s
```

```
time kubectl get --raw /api/v1/nodes/$NODE_NAME/proxy/metrics/resource
real    0m13.784s
user    0m0.063s
sys     0m0.016s
``` 

- I tried running `cri-dockerd` v0.2.6 and don't see any performance improvements, not sure if it's coincidental or specific to OS settings but I will run further tests with v0.2.6 to make sure the upstream changes get included by default.

**Comment by nickvth** [comment]
@kinarashah testing v0.2.6 on Ubuntu 20.04 gives some (not big) performance improvements, I think it’s good to keep in sync with upstream version/changes (bugs/possible cve’s)

**Comment by kinarashah** [comment]
I am able to reproduce the issue using SLES 15 and docker storage driver overlay. I don't see the same issue with overlay2 driver, so I am working with overlay for now. Also able to see the same issue with CentOS and Oracle Linux with overlay driver.

**Comment by kinarashah** [comment]
With cri-dockerd `v0.2.0`, cpu usage is normal but metrics performance is low and with `v0.2.1` cri-dockerd post the PR to calculate metrics in parallel, time to calculate the metrics improves but cpu usage increases drastically. 

Default timeout for metrics-server is 10s, which is why we see the `117 05:33:07.848373       1 scraper.go:140] "Failed to scrape node" err="Get \"https://172.31.4.104:10250/metrics/resource\": context deadline exceeded" node="ip-xxx-xx-xx-xx"` errors. The option to change the timeout will be configurable only with metrics-server `v0.7.x` using `kubelet-request-timeout`.

**Comment by kinarashah** [comment]
I am currently looking into how to resolve the metrics errors without resulting into high cpu usage, will keep this thread updated.

**Comment by kinarashah** [comment]
crictl output: (followed by multiple containers with 0%, ignoring them) 

```
time /usr/local/bin/crictl -r unix:///var/run/dockershim.sock stats

CONTAINER           CPU %               MEM                 DISK                INODES
004ea7aad0adf       2.83                0B                  137.4MB              0
20b49e659a322       2.51                0B                  2.818kB              0
acacd073ad779       0.27                0B                  250B                 0
47331f4695511       0.24                0B                  0B                   0
5da0daa3c88aa       0.21                0B                  0B                   0



real	0m39.025s
user	0m0.026s
sys	0m0.011s
```

Also the endpoint is `dockershim.sock` cuz this is on a node with v1.23.x with `cri-dockerd` enabled.

**Comment by cite** [comment]
* OEL 8.6 with the 5.15 UEK kernel needs `options overlay metacopy=off redirect_dir=off`
* OEL 8.6 with the 4.18 kernel doesn't need any options for "Native Overlay Diff" to work
* CentOS 7 doesn't need and module options for "Native Overlay Diff" to work

However, even with upgraded `cri-dockerd` and `docker` binaries, when deploying 250 containers and enabling the monitoring charts, the metrics requests still overwhelm the `docker` socket:

```time="2022-11-21T10:31:15Z" level=error msg="Cannot connect to the Docker daemon at unix:///var/run/docker.sock. Is the docker daemon running? Failed to get stats from container 9a7780aa05ee553f48a7f204773d445957f5b1fcab132ffc1760baeb486717d1"```

Timings are as described above, and the error is seen on OEL 8.6 with both kernels as well as CentOS 7.

I'm running out of ideas :-(

**Comment by kinarashah** [comment]
@cite @chfrank-cgn I have been debugging metrics collection logic and based on the strace for dockerd, I suspect calculation of this field might be causing the cpu spike. https://github.com/Mirantis/cri-dockerd/blob/v0.2.6/core/stats_linux.go#L78 

I am trying to test it out with different setups, but seeing varying results so I cannot say for sure that's the issue. 
Would you be able to see if this shows any improvement for your test cluster? I have built `kinarashah/rke-tools:v0.1.872` which has `kinarshah/cri-dockerd` v0.2.61 with the updated code. You can replace cri-dockerd directly if that works for you or use the following steps to get the new kubernetes version. 

Steps to get the new version: 
- Rancher -> Home page -> Global Settings (last option on left) -> Edit `rke-metadata-config` 
- Change the `url` to `https://raw.githubusercontent.com/rancher/kontainer-driver-metadata/04d659dd1c100c57ae76942a56bdb55d4a5bb783/data/data.json` 
- Rancher -> Home page -> Cluster Management (2nd option under Global Apps) -> Drivers (3rd option) -> Click on `Refresh Kubernetes Metadata` 
- This would now show `v1.24.6-rancher1-2` under dropdown when you create the new cluster.

**Comment by kinarashah** [comment]
Deleted my older comment on the ask to test with a new image, I saw the issue happen right when I thought things were running fine. Will share again once I re-confirm .

**Comment by kinarashah** [comment]
@cite @chfrank-cgn I have been debugging metrics collection logic and based on the strace for dockerd, I suspect calculation of this field might be causing the cpu spike. https://github.com/Mirantis/cri-dockerd/blob/v0.2.6/core/stats_linux.go#L78 

I am trying to test this out on different setups, but seeing varying results so I cannot confirm if the performance gains I am seeing are actually valid. Would you be able to see if this shows any improvement for your test cluster?

 I have built `kinarashah/rke-tools:v0.1.872` which has `kinarshah/cri-dockerd` v0.2.61 with the updated code. You can replace cri-dockerd directly if that works for you or use the following steps to get the new kubernetes version. 

Steps to get the new version: 
- Rancher -> Home page -> Global Settings (last option on left) -> Edit `rke-metadata-config` 
- Change the `url` to `https://raw.githubusercontent.com/rancher/kontainer-driver-metadata/04d659dd1c100c57ae76942a56bdb55d4a5bb783/data/data.json` 
- Rancher -> Home page -> Cluster Management (2nd option under Global Apps) -> Drivers (3rd option) -> Click on `Refresh Kubernetes Metadata` 
- This would now show `v1.24.6-rancher1-2` under dropdown when you create the new cluster.

**Comment by cite** [comment]
@kinarashah  Thanks for your work! Unfortunately, I don't see any improvements here between 0.2.6 and 0.2.61 - it's about twice as fast as the stock version (0.2.1), but the metrics collection still takes way too long and eventually, the docker socket becomes temporarily unusable.

When comparing https://github.com/kubernetes/kubernetes/blob/2b8076aad0e0b8b63ced05155a4ffbae51a1a61d/pkg/kubelet/dockershim/docker_stats_linux.go#L28 and https://github.com/Mirantis/cri-dockerd/blob/v0.2.6/core/stats_linux.go#L78 I see that the original Kubernetes dockershim implementation is returning FS stats, too. While I can see differences here, I don't really if those could impact the stats collection like that.

However, I put the `dockerd` into debugging mode, in which it logs every GET request. It's then possible to replay any request with curl:

```
Nov 22 10:08:54 muc1tes-k8s-ops-worker-2-1 dockerd[134406]: time="2022-11-22T10:08:54.521431140+01:00" level=debug msg="Calling GET /v1.41/containers/a265a91149d8c35a17a7ecd2ad8da13c241dca7ecea07f135d0f83899c7dea26/stats?one-shot=1&stream=0"

```

would translate to:

```
curl --unix-socket /var/run/docker.sock "http:/v1.41/containers/a265a91149d8c35a17a7ecd2ad8da13c241dca7ecea07f135d0f83899c7dea26/stats?one-shot=1&stream=0"
```

I tested many of those calls, but I wasn't able to find one that's especially slow.

However, when running the `curl` command to gather stats in a loop without any breaks between calls, I noticed that `cri-dockerd` seems to create a huge number of requests to `v1.41/containers/<ID>/json` and `/v1.41/containers/f9a6cdd06539d26df75ddfb7d57553e7a4ace0265ab9c045eed02e888b172b38/stats?one-shot=1&stream=0`. The cluster running 1.23 without `cri-dockerd` however wasn't generating any requestst to `/stats` at all, and there were in general only a small fraction of the requests at all. Is there caching involved in the original Kubernetes implementation?

If a lot less requests are generated with the "old" implementation, perhaps that could explain the performance difference?

**Comment by kinarashah** [comment]
@cite Thanks much for getting back with results quickly! Yeah, I think we're almost looking at the same thing :) 

I saw kubernetes upstream use the same code, but strace for overlay was taking way long to read through the files for rancher/charts so skipping that field helped cpu usage stabilize a bit. It helps me to know we eventually hit the docker limit anyway. 

About the he one-shot requests that you are seeing, they are coming from `getContainerStats` call and the logic is here https://github.com/Mirantis/cri-dockerd/blob/master/vendor/github.com/docker/docker/client/container_stats.go#L30. I tried changing it earlier to disable stream but didn't see any improvements. 

But since I am seeing different results than yours on CentOS, I'll create a new build today with few more changes. I'll also look into the number of requests generated again to see if I missed something here.

**Comment by cite** [comment]
@kinarashah  Thanks for pointing out the relevant code!

I added two logfiles from a docker daemon on debugging loglevel. In both cases, I ran the loop to query the local kubelet for data 10 times. Note how the log from 1.23 doesn't show a single metrics call - I don't really understand what's going on, but that certainly explains the performance difference.

[1.23.log](https://github.com/rancher/rancher/files/10073824/1.23.log)
[1.24.log](https://github.com/rancher/rancher/files/10073825/1.24.log)

@nickvth @chfrank-cgn Are you seeing the same?

**Comment by chfrank-cgn** [comment]
Hi @kinarashah - thanks a lot for the new cri-dockerd! 

I successfully upgraded a 1.23.12 cluster to 1.24.6 using your instructions without CPU spikes or abnormal dockerd behavior - v.0.26.1 seems to solve the upgrade issue when the Native Overlay Diff is false.

I've also checked Prometheus metrics, and they seem to be perfectly fine, too.

Thank you!!

![dashboard-1 24 6](https://user-images.githubusercontent.com/42704306/203531122-d89216ee-75c8-4c3f-9ffc-bb96ac60bdd9.png)

**Comment by cite** [comment]
@kinarashah The same difference in the total amount of queries sent to docker is also seen if we use an 1.23.12 cluster and set it to use `cri-dockerd`: Not a single query for `/v1.4x/containers` without `cri-dockerd`  vs. lots of query when not using the built-in dockershim.

**Comment by kinarashah** [comment]
@chfrank-cgn Thanks for testing it out! This does show the disparity between overlay and overlay2 for WriteableLayer bytes calculation. It is improving overlay performance, I have to look into why we didn't see this earlier. 

@cite Thanks much for sending me the daemon log files, I am still trying to figure out the difference and will keep you updated. I am not seeing any big code changes that can explain this so far but I am going to make it exactly same as internal dockershim, use that as a baseline and see if I can find out why.

**Comment by cite** [comment]
I'm seeing the same slow metrics performance with Ubuntu 22.04 LTS on KVM:

```Client:
 Context:    default
 Debug Mode: false
 Plugins:
  app: Docker App (Docker Inc., v0.9.1-beta3)
  buildx: Docker Buildx (Docker Inc., v0.9.1-docker)
  scan: Docker Scan (Docker Inc., v0.21.0)

Server:
 Containers: 18
  Running: 14
  Paused: 0
  Stopped: 4
 Images: 11
 Server Version: 20.10.21
 Storage Driver: overlay2
  Backing Filesystem: extfs
  Supports d_type: true
  Native Overlay Diff: true
  userxattr: false
 Logging Driver: json-file
 Cgroup Driver: systemd
 Cgroup Version: 2
 Plugins:
  Volume: local
  Network: bridge host ipvlan macvlan null overlay
  Log: awslogs fluentd gcplogs gelf journald json-file local logentries splunk syslog
 Swarm: inactive
 Runtimes: io.containerd.runtime.v1.linux runc io.containerd.runc.v2
 Default Runtime: runc
 Init Binary: docker-init
 containerd version: 770bd0108c32f3fb5c73ae1264f7e503fe7b2661
 runc version: v1.1.4-0-g5fd4c4d
 init version: de40ad0
 Security Options:
  apparmor
  seccomp
   Profile: default
  cgroupns
 Kernel Version: 5.15.0-53-generic
 Operating System: Ubuntu 22.04.1 LTS
 OSType: linux
 Architecture: x86_64
 CPUs: 4
 Total Memory: 15.63GiB
 Name: muc1tes-k8s-ops-worker-2-2
 ID: YFMZ:SBS7:KBWH:NNEI:4GDI:YIO5:R6RJ:3IBY:MUAN:DAN4:MXMS:JCCD
 Docker Root Dir: /var/lib/docker
 Debug Mode: false
 HTTP Proxy: http://muc1tes-proxy-vip.proxy.muc1tes.financial.com:3128
 HTTPS Proxy: http://muc1tes-proxy-vip.proxy.muc1tes.financial.com:3128
 No Proxy: localhost,127.0.0.1,docker.adm.financial.com,docker.financial.com,docker-hosted.adm.financial.com,docker-hosted.financial.com
 Registry: https://index.docker.io/v1/
 Labels:
 Experimental: false
 Insecure Registries:
  127.0.0.0/8
 Live Restore Enabled: false
```

**Comment by kinarashah** [comment]
@cite Without cri-dockerd, kubelet seems to be getting cadvisor stats which is why you don't see the `/stats/` calls to dockerd. I've tried testing with caching the results, but not enough improvement yet. Will keep you posted.

**Comment by phoenix-bjoern** [comment]
This bug makes it impossible to run K8S 1.24+ on RKE1 provisioned clusters. AFAIK there is no workaround for the high CPU usage.
IMHO this issue should get the highest priority if Rancher wants to keep its leading position.

**Comment by immanuelfodor** [comment]
Since cri_dockerd is required for 1.23+ RKE1 clusters, we are basically stuck at 1.22.X if we don't want to give any chance to run into the CPU frying issue. How is it possible that Rancher looks over this?

**Comment by immanuelfodor** [comment]
It looks like the root cause of the situation is that nobody feels like working on the solution is their task? 🤦‍♂️ Oh, come on 😃 https://github.com/Mirantis/cri-dockerd/issues/135#issuecomment-1362056718

**Comment by qeternity** [comment]
Wow, we've just run into this after debugging other suspects first. Upgraded our dev cluster to rancher 2.7.2 / k8s 1.25 on latest ubuntu 20.04 and we're still seeing these spikes every 5-10 seconds.

**Comment by phoenix-bjoern** [comment]
@qeternity can you check the installed Kernel version? We observed the CPU spikes when 5.4.0-132-generic was installed and it disappeared after updating to 5.4.0-135-generic.

**Comment by qeternity** [comment]
Yes we are running 5.4.0-135

I am going to try with 22.04 (5.15) and see if anything improves

**Comment by qeternity** [comment]
Ok so the problem persists with 22.04 and 5.15.0-56-generic

**Comment by phoenix-bjoern** [comment]
@qeternity We noticed that installing Rancher Monitoring has an impact on the CPU usage. Maybe uninstalling it in the affected cluster mitigates the issue for you.

**Comment by horihel** [comment]
> @qeternity We noticed that installing Rancher Monitoring has an impact on the CPU usage. Maybe uninstalling it in the affected cluster mitigates the issue for you.

We've tested that (by disabling the kubelet serviceMonitor) and while it lessens the load on dockerd a bit, it's still a significant load - dockerd still spikes to 200% CPU every few seconds.

We also saw dockerd regularly leaking memory very quickly, resulting in nodes crashing - mainly when lots of pod movement happened (other node draining or a big helm chart installing). After setting a hard memory limit via systemd, we managed to only crash docker, not the whole node - which also revealed that the cause of dockerd misbehaving seems to be cri-dockerd hammering it. Only when the kubelet container is restarted, dockerd stops leaking memory and hogging CPU.

**Comment by qeternity** [comment]
@horihel did you delete the `rancher-monitoring-kubelet` service monitor? I am not able to replicate this improvement

**Comment by qeternity** [comment]
Our dev cluster has become completely unusable. For anyone else who comes across this issue, if you begin to experience this issue, performance will continue to degrade over time. You should migrate away from rancher asap before this happens.

**Comment by immanuelfodor** [comment]
Or stay on an outdated version until a viable upgrade path is available

**Comment by qeternity** [comment]
Yes of course don't upgrade > 1.23 but _if_ you already have, things will degrade within a couple of weeks (caveat: sample size n=1)

**Comment by geoff-carr-bzy** [comment]
> Wow, we've just run into this after debugging other suspects first. Upgraded our dev cluster to rancher 2.7.2 / k8s 1.25 on latest ubuntu 20.04 and we're still seeing these spikes every 5-10 seconds.



> This bug makes it impossible to run K8S 1.24+ on RKE1 provisioned clusters. AFAIK there is no workaround for the high CPU usage. IMHO this issue should get the highest priority if Rancher wants to keep its leading position.

I agree with this, especially since k8s 1.23.x goes EOL 2023-02-28!
https://kubernetes.io/releases/#release-v1-23

Is this confirmed as still an issue > Rancher 2.7.0? There is reference to issues in 2.7.2 above but it is not listed under known issues for 2.7.x that I can see.

**Comment by qeternity** [comment]
Yes, we are running 2.7

**Comment by qeternity** [comment]
So does Rancher intend to ignore this issue? Would be nice to know so that people can make migration plans accordingly.

**Comment by Raboo** [comment]
Well all these issues are pretty much reporting the same error
- https://github.com/rancher/rancher/issues/38018
- https://github.com/rancher/rke/issues/2709
- https://github.com/rancher/rke/issues/2938

Oldest one is reported in September 2021. I assume it needs to be an big enterprise (paying) customer that wants this fixed to get attention. Otherwise I guess that all development efforts go to RKE2 and Rancher UI (or knowing Rancher it might go to developing new products(harverester, elemental etc) that will after 1-2 year be abandoned, who wants to bet against me that harverester will be dead in a 1-2 years?). Rancher has never really been good at supporting products for a long time, they'd rather rewrite stuff or innovate with new products.
But the truly sad thing is that there is no official supported/safe way to migrate/convert a RKE1 cluster into RKE2 so we can get rid of RKE1 once and for all.

Another option would be if the community took upon itself to fix this. It is after all an open source project and so is cri-dockerd.

However in my opinion there should be a big red notice everywhere in ranchers website warning people from upgrading the cluster to k8s v1.24 with RKE1 until it's completely fixed and properly tested, meaning not a empty dev cluster with 2 workloads.

**Comment by horihel** [comment]
I agree that a warning should be in the release notes for both RKE and Rancher.

**Comment by chfrank-cgn** [comment]
Not to downplay the issue, but it does not affect all installations - I have RKE1 1.24.8 (up from 1.24.4) now quite happily running on Ubuntu 20.04 and Docker 20.10

**Comment by qeternity** [comment]
> Not to downplay the issue, but it does not affect all installations - I have RKE1 1.24.8 (up from 1.24.4) now quite happily running on Ubuntu 20.04 and Docker 20.10

How many pods are you running and do you have the metrics chart installed?

**Comment by fe-ax** [comment]
> Not to downplay the issue, but it does not affect all installations - I have RKE1 1.24.8 (up from 1.24.4) now quite happily running on Ubuntu 20.04 and Docker 20.10

We're running these exact versions, too, without issues. We have the Rancher monitoring chart installed and a separate advisor setup running because of the cri-o / docker shim.

I say "without issues", but the hardware underneath the cluster is quite powerful and bare-metal. The docker daemon is using way too much CPU power. In this particular cluster, we have seven nodes running ~500 pods.

**Comment by qeternity** [comment]
> We're running these exact versions, too, without issues.

> The docker daemon seems to be using way too much CPU power.

I'm not sure I follow..._dockerd_ cpu usage is the issue at hand.

**Comment by qeternity** [comment]
I don't think that's the case, these are bare metal 5950X processors. And amongst other things, we run very bursty ci/cd jobs on this cluster and when dockerd spikes to very high cpu alongside existing cpu intensive jobs, we get issues deploying new pods. It's a fairly inhospitable environment to be fair, but we never had issues previously.

**Comment by Raboo** [comment]
Well not to "up-play" this issue in my case the cluster became unavailable from rancher and workloads started dying and crashing due to starvation and timeouts and failing to reschedule. It was troublesome to go revert since the nodes became somewhat unresponsive.
It was quite scary and had the implications that the developers felt that k8s is not that safe with the k8s "all eggs in same basket" (i.e. every app running in k8s and when k8s goes into cascade errors all apps suffer) compared to virtual machines where every application has their own VM.

So yeah a big fat red blinking warning when reading upgrade notes and such would have been great. This issue is known since Q2 2021.

**Comment by geoff-carr-bzy** [comment]
Seems both issues discussed above are now mentioned on the [2.7.1 release notes](https://forums.rancher.com/t/rancher-release-v2-7-1/39886): -

> Known Major Issues
Kubernetes Cluster Distributions:
RKE:
In RKE1 clusters on Kubernetes v1.24, cadvisor container labels and metric series break Monitoring V2 dashboards. See [#38934 1](https://github.com/rancher/rancher/issues/38934#issuecomment-1294585708) for a workaround and more details on the issue.
Rotating encryption keys with a custom encryption provider is not supported. See [#30539](https://github.com/rancher/rancher/issues/30539).
Kubernetes RKE1 1.24.x clusters fail to reach an Active state using Oracle Linux 8.4; high CPU usage is also observed. Cluster could be a fresh installation or upgraded to v1.24.x. See [#38816](https://github.com/rancher/rancher/issues/38816).

We have some money on the table, so I have reached out to our account manager asking what the plan is since we, and many others, are in a holding pattern right now.

If I hear anything I will report back

**Comment by kinarashah** [comment]
Hi all, I want to update you on the current status of this issue. Fwiw, it's prioritized as a critical issue internally and I am actively working on trying to get a working solution for it. I'll see if we can highlight the issue more with a bigger warning in our release notes.

A couple of observations: 
- This issue is not consistently reproducible all the time, I've had better chances of reproducing it on `overlay` vs `overlay2`. Changing a few docker settings has helped tune the performance a little, but it doesn't apply across all systems as seen in the comments above.
- There's more information on things I've tried so far https://github.com/Mirantis/cri-dockerd/issues/135. An important takeaway from there was that metrics collection code in `cri-dockerd` is exactly the same as internal dockershim in kubernetes, so it should behave the same way. But this is a net new change with v1.24 around upstream k8s removing dockershim. Unless remote runtime is enabled for <=v1.23, kubelet uses cadvisor metrics by default which has a much better performance than using docker api to collect metrics which is now getting used with v1.24. 
- This issue is also seen on setups without monitoring enabled, so I agree that disabling monitoring or running your own cadvisor with monitoring enabled would also result into cpu spikes compared to older k8s versions. 

Thanks to @cite, @chfrank-cgn and others who've helped so far in narrowing down the issue so far...I want to be able to resolve this as quickly as I can but given the nature of the issue and its complexity, it'll be most helpful if you can share your setup details and docker info. I'll keep this thread updated when I have more to share, thank you!!

**Comment by geoff-carr-bzy** [comment]
> Seems both issues discussed above are now mentioned on the [2.7.1 release notes](https://forums.rancher.com/t/rancher-release-v2-7-1/39886) so that is a start!: -
> 
> > Known Major Issues
> > Kubernetes Cluster Distributions:
> > RKE:
> > In RKE1 clusters on Kubernetes v1.24, cadvisor container labels and metric series break Monitoring V2 dashboards. See [#38934 1](https://github.com/rancher/rancher/issues/38934#issuecomment-1294585708) for a workaround and more details on the issue.
> > Rotating encryption keys with a custom encryption provider is not supported. See [#30539](https://github.com/rancher/rancher/issues/30539).
> > Kubernetes RKE1 1.24.x clusters fail to reach an Active state using Oracle Linux 8.4; high CPU usage is also observed. Cluster could be a fresh installation or upgraded to v1.24.x. See [#38816](https://github.com/rancher/rancher/issues/38816).
> 
> We have some money on the table, so I have reached out to our account manager asking what the plan is since we, and many others, are in a holding pattern right now.
> 
> If I hear anything I will report back

Had a response from our account manager....they confirmed they have many customers with this issue and a fix is in development but that there is no timeline. This issue has been escalated to have the highest level

**Comment by qeternity** [comment]
@kinarashah fwiw we are only using overlay2 and have experienced this issue on a very vanilla setup.

**Comment by kinarashah** [comment]
@qeternity Could you give me your setup details if possible? Number of nodes, `uname -a` and `docker info`? We haven't been able to able to reproduce this `overlay2` internally so that will be helpful. cc @thaneunsoo

**Comment by qeternity** [comment]
We are running 5 nodes

```
Linux [redacted] 5.4.0-137-generic #154-Ubuntu SMP Thu Jan 5 17:03:22 UTC 2023 x86_64 x86_64 x86_64 GNU/Linux
```

```
Client:
 Context:    default
 Debug Mode: false
 Plugins:
  app: Docker App (Docker Inc., v0.9.1-beta3)
  buildx: Docker Buildx (Docker Inc., v0.9.1-docker)

Server:
 Containers: 261
  Running: 260
  Paused: 0
  Stopped: 1
 Images: 76
 Server Version: 20.10.22
 Storage Driver: overlay2
  Backing Filesystem: extfs
  Supports d_type: true
  Native Overlay Diff: true
  userxattr: false
 Logging Driver: json-file
 Cgroup Driver: cgroupfs
 Cgroup Version: 1
 Plugins:
  Volume: local
  Network: bridge host ipvlan macvlan null overlay
  Log: awslogs fluentd gcplogs gelf journald json-file local logentries splunk syslog
 Swarm: inactive
 Runtimes: io.containerd.runc.v2 io.containerd.runtime.v1.linux runc
 Default Runtime: runc
 Init Binary: docker-init
 containerd version: 5b842e528e99d4d4c1686467debf2bd4b88ecd86
 runc version: v1.1.4-0-g5fd4c4d
 init version: de40ad0
 Security Options:
  apparmor
  seccomp
   Profile: default
 Kernel Version: 5.4.0-137-generic
 Operating System: Ubuntu 20.04.5 LTS
 OSType: linux
 Architecture: x86_64
 CPUs: 32
 Total Memory: 125.6GiB
 Name: prod-node-0
 ID:
 Docker Root Dir: /var/lib/docker
 Debug Mode: false
 Registry: https://index.docker.io/v1/
 Labels:
 Experimental: false
 Insecure Registries:
  127.0.0.0/8
 Live Restore Enabled: false
```

**Comment by georgesouzafarias** [comment]
Same here.

**Comment by kinarashah** [comment]
Hi all, wanted to share another update. The main issue here seems to be Kubernetes deprecating internal dockershim with 1.24 and Rancher now using cri-dockerd - this means that all stats are now collected by cri-dockerd talking to docker api directly instead of getting them from cadvisor. 

I've built a test image that falls back to reading stats from cadvisor, it's showing substantial improvements in the cpu usage. I'm curious to see if it shows any improvement in your setups as well. Note this is not a released version and we're still internally testing it, so we don't recommend using it in production. 

The image is `rancher/hyperkube:v1.24.10-rancher3` and is bundled with `v1.24.10-rancher3-1`. It's available for both 2.6.x and 2.7.x. 

Steps to get the new version:

- Rancher -> Home page -> Global Settings (last option on left) -> Edit `rke-metadata-config`
- Change the url to 
2.6: https://releases.rancher.com/kontainer-driver-metadata/dev-v2.6-0123-patches/data.json
2.7: https://releases.rancher.com/kontainer-driver-metadata/dev-v2.7/data.json 
- Rancher -> Home page -> Cluster Management (2nd option under Global Apps) -> Drivers (3rd option) -> Click on Refresh Kubernetes Metadata
This would now show `v1.24.10-rancher3-1` under dropdown when you create the new cluster.

Thank you!!

**Comment by qeternity** [comment]
Hi @kinarashah thanks very much for the update. Would it be possible to get a 1.25.5+ build. I will happily deploy to our dev cluster for some testing and feedback.

**Comment by snasovich** [comment]
@qeternity , we're working on 1.25.x build; however, please note that any released Rancher version does not yet support k8s 1.25.

**Comment by AcidAngel21** [comment]
Hi @kinarashah, with your test image I have a normal CPU usage in my test cluster. That's awesome. Thanks a lot.
Do you have a timeline when there will be an official release that can be used in production?

**Comment by qeternity** [comment]
@snasovich many thanks, and yes I am aware - early on we attempted a 2.7 deploy to see if that mitigated this cpu usage.

**Comment by orz-nil** [comment]
in my test:
aws ec2 node instance type t3a.medium 
rancher: 2.6.10
os: ubuntu 20.04
kernel: 5.15.0-1019-aws #23~20.04.1-Ubuntu SMP Thu Aug 18 03:20:14 UTC 2022 x86_64 x86_64 x86_64 GNU/Linux
custom rke cluster:
- controlplan: 1
- etcd: 1
- worker: 1
- ingress worker: 1

## the cpu usage grahp
cluster1
<img width="1296" alt="image" src="https://user-images.githubusercontent.com/90177430/219323258-6d22dc60-4fd6-4b11-a8e5-ceee7e273c8a.png">
cluster2
<img width="1333" alt="image" src="https://user-images.githubusercontent.com/90177430/219302368-df418812-f571-4efd-b88a-01c15a93318f.png">

**Comment by chfrank-cgn** [comment]
CPU consumption with v1.24.10-rancher3 is back to normal - thank you!

**Comment by kinarashah** [comment]
@AcidAngel21 @orz-nil @chfrank-cgn Thanks for testing it out!! We're still testing it internally for regressions and such, I'll update you on the official release timeline when I have some concrete info to send out. 

@qeternity Testing for v1.25 might not give you an accurate picture since it's still in development and isn't supported by Rancher yet. We'll try to get it out soon but the same fix exists in v1.24.10-rancher3-1 as well in case you wanted to test v1.24.x out.

**Comment by nickvth** [comment]
Here also CPU consumption with v1.24.10-rancher3 is back to normal on our dev environment. Thanks for this “dirty” k8s version, now looking forward for officieel release.

**Comment by WarpRat** [comment]
Big thanks to everyone involved with working on these fixes! Will this CPU fix also address the [issue](https://github.com/rancher/rancher/issues/38934) with scraping metrics from cadvisor or do we still need to run the workaround with the standalone cadvisor deployment to get properly labeled pod metrics?

**Comment by kinarashah** [comment]
@WarpRat I am not 100% sure if the root cause for both of them are the same even though they look similar. I'll test it out and let you know.

**Comment by AcidAngel21** [comment]
@WarpRat I have tried it out. The workaround is no longer needed.

**Comment by kinarashah** [comment]
Awesome, thank you @AcidAngel21! 

I have also validated all the missing fields mentioned here https://github.com/rancher/rancher/issues/38934#issuecomment-1289428142 and can confirm the issue is resolved. Thanks for calling it out @WarpRat.

**Comment by kinarashah** [comment]
@thaneunsoo For testing, 

High cpu usage on RKE1 clusters >= 1.24 or when enable_cri_dockerd: true for 1.21-1.23 clusters. 
No difference between custom and node provisioned clusters, could use either. 
- top shows high cpu usage 
- one some setups, you might see metrics-server show deadline exceeded errors 

```
E1021 15:52:26.622775       1 scraper.go:140] "Failed to scrape node" 
err="Get \"https://xxx.xx.xx.xxx:10250/metrics/resource\": context deadline exceeded" node="ip-xxx-xx-xx-xxx" 
``` 

this could also be checked by getting stats on the node itself, sys time is significantly high:

1. Don't need monitoring to reproduce 
2. More easier to reproduce using overlay than overlay2, native diff false vs true - we should validate on all settings to make sure the cpu improvement is observed on all setups 
3. Compare output of `stats/summary` between 1.23 and 1.24 to make all the metrics fields are returned
```
curl -sk --key /etc/kubernetes/ssl/kube-apiserver-key.pem --cert /etc/kubernetes/ssl/kube-apiserver.pem 'https://localhost:10250/stats/summary' >/dev/null
```
4. Deploy containers at scale and make sure cpu increase isn't proportional
5. dockerd might show high cpu when pulling images, deploying/deleting containers etc, that's expected

**Comment by cite** [comment]
> * https://releases.rancher.com/kontainer-driver-metadata/dev-v2.7/data.json

@kinarashah I can confirm that we are not seeing any load or other performance problems anymore with `v1.24.10-rancher3-1` - thanks a lot for all your work, and looking forward to the release :-)

**Comment by snasovich** [comment]
Moved this to [v2.7-KDM-For-2.6.11 milestone](https://github.com/rancher/rancher/milestone/309), the move of changes to correct KDM branch is being done per https://github.com/rancher/kontainer-driver-metadata/pull/1089.

**Comment by sowmyav27** [comment]
@thaneunsoo please validate on 2.7.1 Rancher using `KDM branch dev-v2.7-for-2.6.11` for this issue ^

**Comment by cite** [comment]
@kinarashah Will this fix be included in standalone RKE1 clusters, too? If so, do you know which version of RKE1 the changes are targeted for?

**Comment by thaneunsoo** [comment]
### Test Environment: ###
Rancher version: v2.7.1 with KDM `dev-v2.7-for-2.6.11` branch
Rancher cluster type: HA
Docker version: 20.10

Downstream cluster type: rke1

----------

### Testing: ###

Tested this issue with the following steps:
1. Create ami with overlay native diff configurations set
  - set config for overlay and native diff
    - add/change configuration for overlay/overlay2 in `etc/docker/daemon.json`
```
{
  "storage-driver": "overlay2"
}
```

  - restart docker
  - set config for native diff
    - Add the following to `/etc/modprobe.d/disable-overlay-redirect-dir.conf`
```
options overlay metacopy=off redirect_dir=off
```
  - stop docker, reload the module, and restart docker
```
systemctl stop docker
modprobe -r overlay
modprobe overlay
systemctl start docker
```
2. Install KVM ( was unable to reproduce without kvm for some reason)
 - `sudo apt install cpu-checker`
3.Create downstream rke1 cluster using 1.24
4. Create multiple pods (created 100 generic pods)
  - `for i in {1..100}; do kubectl run test$i --image=nginx --port=80; done`
5. Display cpu usage("top")

Side note
- 1.23 cluster was also created to check for differences/similarities
- Used aws node driver and also azure node driver clusters for testing with 3 etcd, 2 cp, 3 worker

Cases covered:
1. overlay with native diff true
2. overlay with native diff false
3. overlay2 with native diff true
4. overlay2 with native diff false

**Result**

**1. overlay with native diff true**
 - before the fix, the cpu usage was ![image.png](https://zube.io/files/rancher/8a0a5831fabb2dd023f8e20b92ca6ac1-image.png)
 - after the fix, the cpu usage is now ![image.png](https://zube.io/files/rancher/bcc873ef17bd02630291fd1ea8dc8be4-image.png)

**2. overlay with native diff false**
 - before the fix the cpu usage was ![image.png](https://zube.io/files/rancher/9d68c890d65c577d5d33d411bfa5867f-image.png)
 - after the fix, the cpu usage is now ![image.png](https://zube.io/files/rancher/8c1cec588b4f057748ba2487c271dd9e-image.png)

**3. overlay2 with native diff true**
 - before the fix the cpu usage was ![image.png](https://zube.io/files/rancher/f62034a1b55145ae4e29d4072dc785c3-image.png)
 - after the fix, the cpu usage is now ![image.png](https://zube.io/files/rancher/cf1409a1f5ce1265ca3b23594dec2fb7-image.png)

**4. overlay2 with native diff false**
 - before the fix the cpu usage was ![image.png](https://zube.io/files/rancher/a5b17ba7665ff1678e727529e069d2fc-image.png)
 - after the fix, the cpu usage is now ![image.png](https://zube.io/files/rancher/83815ee61e9a3290aff5708680da4b94-image.png)

**Comment by kinarashah** [comment]
@thaneunsoo @git-ival 
As discussed offline, I've updated version to `v1.24.10-rancher4-1` in KDM. It doesn't have any new code changes - just build changes to reflect the right k8s version. 

KDM branch: `dev-v2.7-for-2.6.11` / `dev-v2.6` 
RKE RC: `v1.4.3-rc8`/ `v1.3.19-rc3`

**Comment by git-ival** [comment]
## Test Environment:

### Rancher Cluster Config:
Rancher version: v2.7.1 with KDM `dev-v2.7-for-2.6.11` branch
Rancher cluster type: HA , 3 all-roles nodes, 1 worker node (tainted to run rancher-monitoring)
Docker version: 20.10
K8s version: v1.24.10-rancher3-1

### Downstream Cluster Config:
Downstream cluster type: HA RKE1, 5 etcd nodes, 6 control-plane nodes, 10 worker nodes
Installed apps: rancher-monitoring
K8s version: v1.24.10-rancher3-1

### Reproduction Results:
I was able to reproduce on `overlay`, and `overlay2` for pre-fix K8s versions. I saw CPU usage on the downstream worker nodes climb up to the 400%+ range every few minutes. I found it easier to reproduce on `overlay2` with native diff overlay off.

#### Overlay repro:
![Pasted image 20230224195535](https://user-images.githubusercontent.com/85186645/221652482-baa19748-09df-4a6d-bb2f-6f4e3129cc13.png)

#### Overlay2 Native Diff false repro:
![Pasted image 20230227132149](https://user-images.githubusercontent.com/85186645/221652443-17ed4fa5-b912-4ff5-afd6-4bdbddc0fa88.png)

#### Overlay2 Native Diff true repro:
![Pasted image 20230224195900](https://user-images.githubusercontent.com/85186645/221652269-b7b6f5cc-3ae5-41aa-a1f0-087270198453.png)

### Test Steps:
1. Create Rancher cluster
	- Ensure the KDM branch is set to the `dev-v2.7-for-2.6.11` branch by setting the `rke-metadata-config` Global Setting url value
2. Create Downstream cluster (1 for each case below)
	- Create NodeTemplates for each Downstream cluster
		- Set the Engine Option for the Storage Driver as appropriate for the current test case
3. Wait for Downstream cluster to provision
4. (Overlay2 only) Set Native Overlay Diff on each cluster node as needed: 
	1. Stop docker
	2. Reload the modprobe modules
	3. Set modprobe options for Native Diff
	4. Start docker
`systemctl stop docker && modprobe -r overlay && modprobe overlay metacopy=off redirect_dir=off && systemctl start docker`
6. Scale-in a large # of pods (created 900 nginx Deployments with 1 pod each)
7. SSH into a control-plane and worker node
8. Run `top` to monitor CPU usage as scale-in proceeds
9. Once scale-in completes, proceed to Grafana metrics to confirm resource usager %s
   
Notes:
- Provisioned all clusters via Terraform
- Downstreams were all AWS nodedriver clusters
- Used grafana metrics for results screenshots since the `top` screenshots are less useful

Cases covered:
1. Overlay
2. Overlay2 with native diff true
3. Overlay2 with native diff false

## Test Case Results:

### Overlay - Test Case Results:

#### Pre Pod Scale-in Phase:
- 17% CPU usage
- 6% Memory Usage

#### Post Pod Scale-in:
- CPU Usage never reaches > 77%
- Memory Usage never reaches > 47%

![Pasted image 20230224003826](https://user-images.githubusercontent.com/85186645/221652218-41cf01e9-eebb-478c-9d5f-f9231423afe3.png)
![Pasted image 20230224004900](https://user-images.githubusercontent.com/85186645/221652191-8e279e18-5d53-4bc3-bea5-ff438f1f990b.png)

### Overlay2 Native Diff false -  Test Case Results:

#### Pre Pod Scale-in Phase:
- ~13% CPU usage
-  ~4% Memory Usage

#### Post Pod Scale-in:
- CPU Usage never reaches > 60%
- Memory Usage never reaches > 51%

![Pasted image 20230224004415](https://user-images.githubusercontent.com/85186645/221652166-9ca29ed5-30e7-4283-ab2a-ebfd70218a4a.png)
![Pasted image 20230224004919](https://user-images.githubusercontent.com/85186645/221652140-3ea67ffd-a51f-4b8f-8ed2-c282637c8998.png)


### Overlay2 Native Diff true - Test Case Results:

#### Pre Pod Scale-in Phase:
- ~12% CPU usage
- ~3% Memory Usage

#### Post Pod Scale-in:
- CPU Usage never reaches > 70%
- Memory Usage never reaches > 53%

![Pasted image 20230224004433](https://user-images.githubusercontent.com/85186645/221652046-1840a742-5895-4eab-a502-51a9bd33da4a.png)
![Pasted image 20230224004936](https://user-images.githubusercontent.com/85186645/221652018-b4e9e5a4-76bf-4350-85b3-bc2abd6b2168.png)

**Comment by git-ival** [comment]
Validations completed, closing.

**Comment by qeternity** [comment]
Hi @kinarashah thanks so much for your hard work. I know it's not GA but we've deployed 1.25.6-rancher4-1 to our dev cluster and have seen massive reduction in CPU usage.

**Comment by AcidAngel21** [comment]
@kinarashah `v1.24.10-rancher4-1` is shown under dropdown.
Is this an official release that I can use in production?
![image](https://user-images.githubusercontent.com/64893006/223425528-ac19c2a9-cf7d-43b3-97b7-a5fd2cdf24f7.png)

**Comment by immanuelfodor** [comment]
Can confirm the CPU issue fixed: upgraded v1.23.16-rancher2-1 -> v1.24.13-rancher2-1 with `enable_cri_dockerd: true` in RKE1 and the CPU usage is normal.
